packages/labs/signals/src/lib/signal-watcher.ts

```diff
@@ -87,25 +177,70 @@ export function SignalWatcher<T extends ReactiveElementConstructor>(
           return;
         }
         if (el.__forcingUpdate === false) {
-          el.requestUpdate();
+          const needsUpdate = new Set(this.getPending()).has(
+            el.__performUpdateSignal as Signal.Computed<void>
+          );
+          if (needsUpdate) {
+            el.requestUpdate();
+          } else {
+            el.__queueEffects();
+          }
         }
         this.watch();
       }));
       elementForWatcher.set(watcher, this as unknown as SignalWatcherInternal);
-      elementFinalizationRegistry.register(this, {
-        watcher,
-        signal: this.__performUpdateSignal,
-      });
+      elementFinalizationRegistry.register(this, watcher);
       watcher.watch(this.__performUpdateSignal);
+      watcher.watch(...Array.from(this.__effects).map(([signal]) => signal));
     }
 
     private __unwatch() {
-      if (this.__watcher === undefined) {
+      if (this._watcher === undefined) {
         return;
       }
-      this.__watcher.unwatch(this.__performUpdateSignal!);
+      // We unwatch all signals that are not manually disposed, so that we don't
+      // keep the element alive by holding references to it.
+      this._watcher.unwatch(
```

Comment:
Do we do this manually because of manualDispose, or is there also no bulk way to dispose of the whole watcher?